File: engine-python/src/main/java/pl/psi/aaas/engine/python/GrpcEngine.java
package pl.psi.aaas.engine.python;

import pl.psi.aaas.Engine;
import pl.psi.aaas.Parameter;
import pl.psi.aaas.Vector;
import pl.psi.aaas.communication.*;
import pl.psi.aaas.usecase.CalculationDefinition;
import pl.psi.aaas.usecase.CalculationException;
import pl.psi.aaas.usecase.Parameters;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Map;
import java.util.stream.Collectors;

public class GrpcEngine implements Engine<CalculationDefinition, Parameters, Parameters> {

    private final ConnectionProvider<AAASGrpc.AAASBlockingStub> connectionProvider;

    public GrpcEngine(ConnectionProvider<AAASGrpc.AAASBlockingStub> connectionProvider) {
        this.connectionProvider = connectionProvider;
    }

    @Override
    public Parameters call(CalculationDefinition calcDef) {
        Map<String, pl.psi.aaas.communication.Parameter> mappedParams = calcDef.getInParameters().entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, entry -> toCommParams(entry.getValue())));

        pl.psi.aaas.communication.Parameters paramsToBeSent = pl.psi.aaas.communication.Parameters.newBuilder()
                .putAllValue(mappedParams)
                .build();

        AAASGrpc.AAASBlockingStub client = connectionProvider.getConnection();

        CallParameters callParams = CallParameters.newBuilder()
                .setParams(paramsToBeSent)
                .setScript(calcDef.getCalculationScript())
                .build();
        try {
            CallParametersResponse retParams = client.call(callParams);
            if (!retParams.getErrorMessage().isEmpty()) {
                throw new CalculationException(retParams.getErrorMessage());
            }
            Map<String, Parameter<?>> resultParams = retParams.getParams().getValueMap().entrySet().stream()
                    .collect(Collectors.toMap(
                            Map.Entry::getKey,
                            entry -> fromCommParams(entry.getValue().getVector(), calcDef.getOutParameters().get(entry.getKey()))
                    ));
            return new Parameters(resultParams);
        } catch (Exception ex) {
            throw new CalculationException(ex.getMessage() != null ? ex.getMessage() : "Unknown error during calculation", ex);
        }
    }

    public pl.psi.aaas.communication.Parameter toCommParams(Parameter<?> value) {
        if (!(value instanceof Vector)) {
            throw new CalculationException("Unknown value type " + value.getClazz());
        }
        Vector<?> vector = (Vector<?>) value;
        pl.psi.aaas.communication.Vector commVector = toCommVector(vector);

        return pl.psi.aaas.communication.Parameter.newBuilder()
                .setVector(commVector)
                .build();
    }

    private pl.psi.aaas.communication.Vector toCommVector(Vector<?> vector) {
        pl.psi.aaas.communication.Vector.Builder vectorBuilder = pl.psi.aaas.communication.Vector.newBuilder();
        Class<?> elemClazz = vector.getElemClazz();

        if (elemClazz == Double.class || elemClazz == double.class) {
            toDoubleCommVector((Vector<Double>) vector, vectorBuilder);
        } else if (elemClazz == Long.class || elemClazz == long.class) {
            toLongCommVector((Vector<Long>) vector, vectorBuilder);
        } else if (elemClazz == String.class) {
            toStringCommVector((Vector<String>) vector, vectorBuilder);
        } else if (elemClazz == Boolean.class || elemClazz == boolean.class) {
            toBooleanCommVector((Vector<Boolean>) vector, vectorBuilder);
        } else if (elemClazz == Vector.class) {
            toVectorCommVector((Vector<Vector<?>>) vector, vectorBuilder);
        } else if (elemClazz == ZonedDateTime.class) {
            toZonedDateTimeCommVector((Vector<ZonedDateTime>) vector, vectorBuilder);
        } else {
            throw new CalculationException("Unsupported type: " + elemClazz);
        }
        return vectorBuilder.build();
    }

    private void toDoubleCommVector(Vector<Double> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<DoubleValue> iterable = vector.getValue().stream().map(it -> {
            DoubleValue.Builder builder = DoubleValue.newBuilder();
            if (it == null) {
                builder.setEmpty(EmptyValue.getDefaultInstance());
            } else {
                builder.setValue(it);
            }
            return builder.build();
        }).collect(Collectors.toList());
        vectorBuilder.setDoubleValue(DoubleVector.newBuilder().addAllVector(iterable).build());
    }

    private void toLongCommVector(Vector<Long> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<LongValue> iterable = vector.getValue().stream().map(it -> {
            LongValue.Builder builder = LongValue.newBuilder();
            if (it == null) {
                builder.setEmpty(EmptyValue.getDefaultInstance());
            } else {
                builder.setValue(it);
            }
            return builder.build();
        }).collect(Collectors.toList());
        vectorBuilder.setLongValue(LongVector.newBuilder().addAllVector(iterable).build());
    }

    private void toStringCommVector(Vector<String> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<StringValue> iterable = vector.getValue().stream().map(it -> {
            StringValue.Builder builder = StringValue.newBuilder();
            if (it == null) {
                builder.setEmpty(EmptyValue.getDefaultInstance());
            } else {
                builder.setValue(it);
            }
            return builder.build();
        }).collect(Collectors.toList());
        vectorBuilder.setStringValue(StringVector.newBuilder().addAllVector(iterable).build());
    }

    private void toBooleanCommVector(Vector<Boolean> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<BoolValue> iterable = vector.getValue().stream().map(it -> {
            BoolValue.Builder builder = BoolValue.newBuilder();
            if (it == null) {
                builder.setEmpty(EmptyValue.getDefaultInstance());
            } else {
                builder.setValue(it);
            }
            return builder.build();
        }).collect(Collectors.toList());
        vectorBuilder.setBoolValue(BoolVector.newBuilder().addAllVector(iterable).build());
    }

    private void toZonedDateTimeCommVector(Vector<ZonedDateTime> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<TimestampValue> iterable = vector.getValue().stream().map(it -> {
            TimestampValue.Builder builder = TimestampValue.newBuilder();
            if (it == null) {
                builder.setEmpty(EmptyValue.getDefaultInstance());
            } else {
                builder.setValue(it.toInstant().toEpochMilli());
            }
            return builder.build();
        }).collect(Collectors.toList());
        vectorBuilder.setTimestampValue(TimestampVector.newBuilder().addAllVector(iterable).build());
    }

    private void toVectorCommVector(Vector<Vector<?>> vector, pl.psi.aaas.communication.Vector.Builder vectorBuilder) {
        Iterable<pl.psi.aaas.communication.Vector> iterable = vector.getValue().stream().map(it -> {
            if (it == null) {
                return pl.psi.aaas.communication.Vector.newBuilder().build();
            } else {
                return toCommVector(it);
            }
        }).collect(Collectors.toList());
        vectorBuilder.setVectorValue(VectorVector.newBuilder().addAllVector(iterable).build());
    }

    public Vector<?> fromCommParams(pl.psi.aaas.communication.Vector vector, Parameter<?> parameter) {
        if (parameter instanceof Vector) {
            if (vector.hasDoubleValue()) return fromCommDoubleVector(vector);
            if (vector.hasLongValue()) return fromCommLongVector(vector);
            if (vector.hasBoolValue()) return fromCommBoolVector(vector);
            if (vector.hasStringValue()) return fromCommStringVector(vector);
            if (vector.hasTimestampValue()) return fromCommTimestampVector(vector);
            if (vector.hasEmptyValue()) return (Vector<?>) parameter;
            // TODO: handle vector value if needed
            throw new CalculationException("Unsupported vector type: " + vector);
        } else {
            throw new CalculationException("Unsupported operation, parameter is null or it's not a Vector<?> ");
        }
    }

    private Vector<ZonedDateTime> fromCommTimestampVector(pl.psi.aaas.communication.Vector vector) {
        ZonedDateTime[] array = vector.getTimestampValue().getVectorList().stream().map(it -> {
            if (it.hasEmpty()) return null;
            return ZonedDateTime.ofInstant(Instant.ofEpochMilli(it.getValue()), ZoneOffset.UTC);
        }).toArray(ZonedDateTime[]::new);
        return Parameter.ofArray(array);
    }

    private Vector<String> fromCommStringVector(pl.psi.aaas.communication.Vector vector) {
        String[] array = vector.getStringValue().getVectorList().stream().map(it -> {
            if (it.hasEmpty()) return null;
            return it.getValue();
        }).toArray(String[]::new);
        return Parameter.ofArray(array);
    }

    private Vector<Boolean> fromCommBoolVector(pl.psi.aaas.communication.Vector vector) {
        Boolean[] array = vector.getBoolValue().getVectorList().stream().map(it -> {
            if (it.hasEmpty()) return null;
            return it.getValue();
        }).toArray(Boolean[]::new);
        return Parameter.ofArray(array);
    }

    private Vector<Long> fromCommLongVector(pl.psi.aaas.communication.Vector vector) {
        Long[] array = vector.getLongValue().getVectorList().stream().map(it -> {
            if (it.hasEmpty()) return null;
            return it.getValue();
        }).toArray(Long[]::new);
        return Parameter.ofArray(array);
    }

    private Vector<Double> fromCommDoubleVector(pl.psi.aaas.communication.Vector vector) {
        Double[] array = vector.getDoubleValue().getVectorList().stream().map(it -> {
            if (it.hasEmpty()) return null;
            return it.getValue();
        }).toArray(Double[]::new);
        return Parameter.ofArray(array);
    }
}

File: engine-python/src/main/kotlin/pl/psi/aaas/engine/python/GrpcEngine.kt
package pl.psi.aaas.engine.python

import pl.psi.aaas.Engine
import pl.psi.aaas.Parameter
import pl.psi.aaas.Vector
import pl.psi.aaas.communication.*
import pl.psi.aaas.usecase.CalculationDefinition
import pl.psi.aaas.usecase.CalculationException
import pl.psi.aaas.usecase.Parameters
import java.time.Instant
import java.time.ZoneOffset
import java.time.ZonedDateTime
import pl.psi.aaas.communication.Parameter as ParameterComm
import pl.psi.aaas.communication.Parameters as ParametersComm
import pl.psi.aaas.communication.Vector as VectorComm

class GrpcEngine(val connectionProvider: ConnectionProvider<AAASGrpc.AAASBlockingStub>) : Engine<CalculationDefinition, Parameters, Parameters> {

    override fun call(calcDef: CalculationDefinition): Parameters? {
        val mappedParams = calcDef.inParameters.map { it.key to toCommParams(it.value) }.toMap()
        val paramsToBeSent = ParametersComm.newBuilder()
                .putAllValue(mappedParams)
                .build()

        val client = connectionProvider.getConnection()

        val callParams = CallParameters.newBuilder()
                .setParams(paramsToBeSent)
                .setScript(calcDef.calculationScript)
                .build()
        try {
            val retParams = client.call(callParams)
            if (retParams.errorMessage.isNotEmpty())
                throw CalculationException(retParams.errorMessage)
            return retParams.params.valueMap.map { it.key to fromCommParams(it.value.vector) }.toMap()
        } catch (ex: Exception) {
            throw CalculationException(ex.message ?: "Unknown error during calculation", ex)
        }
    }


    fun toCommParams(value: Parameter<*>): ParameterComm {
        val vector = when (value) {
            is Vector<*> -> toCommVector(value)
            else         -> throw CalculationException("Unknown value type ${value.clazz}")
        }

        val paramBuilder = ParameterComm.newBuilder()
        return paramBuilder.setVector(vector).build()
    }

    private fun toCommVector(vector: Vector<*>): VectorComm {
        val vectorBuilder = VectorComm.newBuilder()
        when (vector.elemClazz) {
            Double::class.java            -> toDoubleCommVector(vector, vectorBuilder)
            java.lang.Double::class.java  -> toDoubleCommVector(vector, vectorBuilder)
            Long::class.java              -> toLongCommVector(vector, vectorBuilder)
            java.lang.Long::class.java    -> toLongCommVector(vector, vectorBuilder)
            String::class.java            -> toStringCommVector(vector, vectorBuilder)
            java.lang.String::class.java  -> toStringCommVector(vector, vectorBuilder)
            Boolean::class.java           -> toBooleanCommVector(vector, vectorBuilder)
            java.lang.Boolean::class.java -> toBooleanCommVector(vector, vectorBuilder)
            Vector::class.java            -> toVectorCommVector(vector, vectorBuilder)
            ZonedDateTime::class.java     -> toZonedDateTimeCommVector(vector, vectorBuilder)
            else                          -> throw CalculationException("Unsupported type: " + vector.elemClazz)
        }
        return vectorBuilder.build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toDoubleCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<Double?>
        val iterable = vector.value.map {
            with(DoubleValue.newBuilder()) {
                if (it == null)
                    setEmpty(EmptyValue.getDefaultInstance())
                else
                    setValue(it)
            }.build()
        }.asIterable()
        vectorBuilder.doubleValue = DoubleVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toLongCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<Long?>
        val iterable = vector.value.map {
            with(LongValue.newBuilder()) {
                if (it == null)
                    setEmpty(EmptyValue.getDefaultInstance())
                else
                    setValue(it)
            }.build()
        }.asIterable()
        vectorBuilder.longValue = LongVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toStringCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<String?>
        val iterable = vector.value.map {
            with(StringValue.newBuilder()) {
                if (it == null)
                    setEmpty(EmptyValue.getDefaultInstance())
                else
                    setValue(it)
            }.build()
        }.asIterable()
        vectorBuilder.stringValue = StringVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toBooleanCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<Boolean?>
        val iterable = vector.value.map {
            with(BoolValue.newBuilder()) {
                if (it == null)
                    setEmpty(EmptyValue.getDefaultInstance())
                else
                    setValue(it)
            }.build()
        }.asIterable()
        vectorBuilder.boolValue = BoolVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toZonedDateTimeCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<ZonedDateTime?>
        val iterable = vector.value.map {
            with(TimestampValue.newBuilder()) {
                if (it == null)
                    setEmpty(EmptyValue.getDefaultInstance())
                else
                    setValue(it.toInstant().toEpochMilli())
            }.build()
        }.asIterable()
        vectorBuilder.timestampValue = TimestampVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toVectorCommVector(vector: Vector<*>, vectorBuilder: VectorComm.Builder) {
        vector as Vector<Vector<Any>?>
        val iterable = vector.value.map {
            if (it == null)
                VectorComm.newBuilder().build()
            else
                toCommVector(it)
        }.asIterable()
        vectorBuilder.vectorValue = VectorVector.newBuilder().addAllVector(iterable).build()
    }

    @Suppress("UNCHECKED_CAST")
    fun fromCommParams(vector: VectorComm): Vector<Any> =
            if (vector.hasDoubleValue()) fromCommDoubleVector(vector)
            else if (vector.hasLongValue()) fromCommLongVector(vector)
            else if (vector.hasBoolValue()) fromCommBoolVector(vector)
            else if (vector.hasStringValue()) fromCommStringVector(vector)
            else if (vector.hasTimestampValue()) fromCommTimestampVector(vector)
            else if (vector.hasVectorValue()) {
                val array = vector.vectorValue.vectorList.map { fromCommParams(it) }.toTypedArray()
                Parameter.ofArray(array) as Vector<Any>
            } else
                throw CalculationException("Unsupported vector type: " + vector)

    @Suppress("UNCHECKED_CAST")
    private fun fromCommTimestampVector(vector: VectorComm): Vector<Any> {
        val array = vector.timestampValue.vectorList.map {
            if (it.hasEmpty())
                null
            else
                ZonedDateTime.ofInstant(Instant.ofEpochMilli(it.value), ZoneOffset.UTC)
        }.toTypedArray()
        return Parameter.ofArray(array) as Vector<Any>
    }

    @Suppress("UNCHECKED_CAST")
    private fun fromCommStringVector(vector: VectorComm): Vector<Any> {
        val array = vector.stringValue.vectorList.map {
            if (it.hasEmpty())
                null
            else
                it.value
        }.toTypedArray()
        return Parameter.ofArray(array) as Vector<Any>
    }

    @Suppress("UNCHECKED_CAST")
    private fun fromCommBoolVector(vector: VectorComm): Vector<Any> {
        val array = vector.boolValue.vectorList.map {
            if (it.hasEmpty())
                null
            else
                it.value
        }.toTypedArray()
        return Parameter.ofArray(array) as Vector<Any>
    }

    @Suppress("UNCHECKED_CAST")
    private fun fromCommLongVector(vector: VectorComm): Vector<Any> {
        val array = vector.longValue.vectorList.map {
            if (it.hasEmpty())
                null
            else
                it.value
        }.toTypedArray()
        return Parameter.ofArray(array) as Vector<Any>
    }

    @Suppress("UNCHECKED_CAST")
    private fun fromCommDoubleVector(vector: pl.psi.aaas.communication.Vector): Vector<Any> {
        val array = vector.doubleValue.vectorList.map {
            if (it.hasEmpty())
                null
            else
                it.value
        }.toTypedArray()
        return Parameter.ofArray(array) as Vector<Any>
    }
}
